### ISSUE TITLE
Restricted registration

### ISSUE THREAD

**Issue by runbeyond** [spectrumtrail/spectrum#141]
I want to be able to restrict registration to a select group of people for a particular event (or race) so that, for example, I can offer advanced registration to the previous year’s participants for a particular event. 

More specific use case: each year, for Sky Island registration, we offer a 48 hour advanced registration window for all participants in the previous year’s event.

**Comment by runbeyond** [comment]
@ColinW520 Please let me know if you have any questions that could help with an implementation solution. This one does not have to be fancy.

**Comment by runbeyond** [comment]
@ColinW520 Do we need to keep this issue?

**Comment by runbeyond** [comment]
Accidentally closed it

**Comment by ColinW520** [comment]
Yes we should keep this issue because it’s a consistent feature needed

**Comment by willblake01** [comment]
@ColinW520 @runbeyond quick question for this. Are you wanting to be able to manually choose a group (not necessarily only people who registered for the same race the previous year) or will the only use case for now be the previous year's participants for a given race?

**Comment by willblake01** [comment]
Still working this out in my head but here's where I am currently.

Steps:

App => models => Add special_registration.rb
App => models => Add special_registration_code.rb

App => services => new_registration_handler.rb => Add create_new_special_registration class 
App => services => Add validates_special_registration_usage

App => controllers => registrations_controller.rb => Add 							 	create_new_special_registration(handler) class
App => controllers => Add special_registration_codes_controller.rb

Migration => Add Special Registrations to DB

App => views => admin => Add special_registration_codes directory similar to discount_codes
	Create Special Registration
	-Code Text (i.e. Sky Island early registration, Sky Island full race addition)
	-Race (that code applies)
	-Limit (how many times can be used)
	-is active boolean (to shut down at any time)
	-valid emails
	-Special Registration Start Date
	-Special Registration End Date

**Comment by runbeyond** [comment]
@willblake01 Thanks for working on this sense! I follow the logic and I think you have a good grasp on the requirements. For everyone's benefits, I will write more detail and acceptance criteria in the story to make sure we're all having the same conversation.

I look forward to Colin's feedback on your implementation approach thus far! Great work and thanks again.

**Comment by ColinW520** [comment]
WOAH.

Very well written plan, @willblake01 !!

How can we implement this _incrementally_?
